### DISCUSSION TITLE
What might cause 503 error

### DISCUSSION THREAD

**Discussion by habibali95** [eclipse-theia/theia#8945] [General]
I'm deploying Theia using Kubernetes, randomly I get the error 503 after the readiness probe passed with success. but after a refresh after a few seconds sometimes more than 5 the 503 disappear.
This is the current configuration
readinessProbe:
tcpSocket:
port: 3000
successThreshold: 10
failureThreshold: 20
periodSeconds: 1
I tried other configs such as httpGet.
does importing the project has to do with the 503? knowing that the project size might be higher than 1GB.
Any help would be really apperciated.

**Comment by marcdumais-work** [comment]
Hi @habibali95 ,

According to the  "Google - I am feeling lucky" technique, `503` means: 

_503 Service Unavailable. The server is currently unable to handle the request due to a temporary overloading or maintenance of the server. The implication is that this is a temporary condition which will be alleviated after some delay. If known, the length of the delay MAY be indicated in a Retry-After header._

I do not know that Theia would be involved in emitting this error, but the Node-builtin HTTP server might? Maybe e.g. if unable to allocate resources to do its work. 

How do you import your project? Git clone? Doing heavy processing in a workspace container, I have sometimes seen momentary disconnections, reported in the frontend app by a change of color of the status bar, from blue to orange. If importing your project or something else stresses the capabilities of your container (CPU, RAM), it's to be expected that it might disrupt the Theia application too.

**Comment by habibali95** [comment]
Hello @marcdumais-work ,

Thank you for your reply, actually, the 503 appears only in the first attempt to open Theia after it's deployed and the readiness probe has succeeded, after refreshing the page Theia is opened normally, I'm using Nginx Ingress to access to Theia and a volumeMount with mountPath of the project. I'm sure the issue has nothing to do with Theia directly but my question is why after the readiness probe returned success, Theia is still not ready and might need more few seconds.. is there any other efficient way to test if Theia is fully up and ready to be used?

Thank you very much!

**Reply by marcdumais-work** [reply]
Hi @habibali95 

Thanks for the extra info. I'm not an expert on Kubernetes so I might be missing something. Others that know more can chime-in.

> Thank you for your reply, actually, the 503 appears only in the first attempt to open Theia after it's deployed and the readiness probe has succeeded

Interesting. This could point to something in your setup happening as you access the app the first time, that does not quite have time to finish, in time, but sticks after. 

> my question is why after the readiness probe returned success, Theia is still not ready and might need more few seconds.. is there any other efficient way to test if Theia is fully up and ready to be used?

Can you share how the probe works? Does it try to access the Theia app port to see if it's responding or e.g. is it checking that the container has started? If the later, something else in your container startup could be delaying the Theia app from starting, such that it has not yet "started to start", when you try to connect? 

When playing locally with the example browser app from our main repo, I never get a 503 - if the Theia app backend is still starting when I try to access, the browser just happily waits until it's ready, and proceeds to load the app.

**Comment by habibali95** [comment]
Hello @marcdumais-work,

Thanks for replying, The kubernetes readiness probe work as the configuration I mentioned above so basically it tries to open a tcp socket on the port 3000, if the connection is opened, the container is considered as healthy and ready to go. But I tried also readiness probe with httpGet instead of tcpSocket, this time Kubernetes will send http get request on port 3000, if the returned status is 2xx, Theia will be considered healthy and ready to be used and still I got the same 503 issue...

Isn't there a health check method implemented in Theia and if not is it possible for me to add it and build a new image?

**Reply by marcdumais-work** [reply]
Does the probe access the app the same way you do as a user or does it get "backdoor access", e.g. accessing the app on an internal network, e.g. bypassing a reverse proxy you might have in front of your containers?

**Comment by habibali95** [comment]
Hello @marcdumais-work, 
I think Kubernetes is accessing the app on an internal network meaning  it will bypass a revere proxy or anything similar

**Reply by marcdumais-work** [reply]
That could be the issue. Maybe there's a way to customize the probe, make it use the external network? The Theia app itself would be unaware whether the network magic to make it accessible is done.

**Reply by marcdumais-work** [reply]
i.e. the probe will currently conclude that the Theia app is almost instantly ready, but the (/your) machinery in front might not be.

**Reply by habibali95** [reply]
I see, thanks a lot for your help @marcdumais-work
